src/harness/database.py

```diff
@@ -0,0 +1,159 @@
+    when the thread is started using thread.start().
+    """
+
+    # Any exceptions in the server should be caught and handled without
+    # intervention and not cause it to stop. This is an added safeguard so
+    # any unhandled errors will result in:
+    #   1) An error being logged (Most important to diagnose problem).
+    #   2) An attempt to restart the server (allowing the test to proceed).
+    self.stopped = False
+    while not self.stopped:
+      logging.info('Started Database Server: %s at %s', self.name,
+                   self.address)
+      try:
+        self.server.serve_forever()
+      except:
+        traceback.print_exc()
+        logging.error('HTTPServer %s at %s died on %s, restarting...',
+                      self.name, self.address,
+                      datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'))
+    logging.info('Shutting down Database Server:%s at %s', self.name,
+                 self.address)
+
+  def setFileToServe(self, file_url, file_path):
+    """Remove existing path mappings and set a single file url to path mapping.
+
+    Args:
+      file_url: The url to provide the file on.
+      file_path: The file path to provide on the given url.
+    """
+    self.server.file_paths = {file_url: file_path}
+
+  def setFilesToServe(self, file_url_file_path_dict):
+    """Remove existing path mappings and set multiple file url to path mappings.
+
+    Args:
+      file_url_file_path_dict: A dictionary of file_url: file_path as described
+        in the setFileToServe description.
+    """
+    self.server.file_paths = file_url_file_path_dict
+
+class DatabaseHTTPServer(HTTPServer):
+  def __init__(self, server_address, RequestHandlerClass, name, authorization):
+    HTTPServer.__init__(self, server_address, RequestHandlerClass)
+    self.name = name
+    self.file_paths = {}
+    # Encode the username and password for simple comparison. Expected form is:
+    #   'Basic Base64RepresentationOfUsername:Password'
+    self.authorization = 'Basic ' + base64.b64encode(authorization) if authorization else None
+
+class DatabaseHandler(SimpleHTTPRequestHandler):
+  def translate_path(self, path):
+    """Translate URL paths into filepaths."""
+    if self.server.authorization:
+      # Failed Authorization will return a 401 error and the reason why.
+      if 'Authorization' not in self.headers:
+        logging.info('Authorization Failed. No Auth header.')
+        self.send_response(401)
+        self.end_headers()
+        self.wfile.write('no auth header received.\n')
+        self.wfile.write('Ignore 404 response below.\n\n')
+        return ""
+      elif self.headers['Authorization'] != self.server.authorization:
+        logging.info('Authorization Failed. Incorrect username:password.')
+        self.send_response(401)
+        self.end_headers()
+        self.wfile.write('Incorrect username:password.\n')
+        self.wfile.write('Ignore 404 response below.\n\n')
+        return ""
```

Comment:
To preempt a reasonably likely comment. Yes, this is a bit of a hacky way to send a 401 response (as it will also return the text for a 404 in the message body).

I do not think this is a problem given that:
- Authentication is only used for one database.
- The tests expect the authentication to work anyway, this code path is only for debugging purposes.
- The amount of work required to get this functioning exactly the same as a real server is significantly more than we would get out of it doing so.
- This server is only for testing.